src/main/java/io/cdap/plugin/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java

```diff
@@ -129,20 +129,95 @@ protected void customize(Request exchange) {
 
     // Now set up the Bayeux client itself
     BayeuxClient client = new BayeuxClient(oAuthInfo.getInstanceURL() + DEFAULT_PUSH_ENDPOINT, transport);
-    client.handshake();
-
     return client;
   }
 
-  private void waitForHandshake(BayeuxClient client,
-                                long timeoutInMilliseconds, long intervalInMilliseconds) {
+  public void createSalesforceListener() {
+    try {
+      bayeuxClient = getClient(credentials);
+    } catch (Exception exception) {
+      LOG.error("Exception while creating bayeux client ", exception);
+    }
+    bayeuxClient.getChannel(Channel.META_HANDSHAKE).addListener
+      ((ClientSessionChannel.MessageListener) (channel, message) -> {
+
+        boolean success = message.isSuccessful();
+        if (!success) {
+          String error = (String) message.get("error");
+          if (error != null) {
+            LOG.error("Error during HANDSHAKE: {}", error);
+          }
+
+          Exception exception = (Exception) message.get("exception");
+          if (exception != null) {
+            LOG.error("Exception during HANDSHAKE", exception);
+          }
+        }
+      });
+
+    bayeuxClient.getChannel(Channel.META_CONNECT).addListener(
+      (ClientSessionChannel.MessageListener) (channel, message) -> {
+
+        boolean success = message.isSuccessful();
+        if (!success) {
+          String error = (String) message.get("error");
+          Map<String, Object> advice = message.getAdvice();
+
+          if (error != null) {
+            LOG.error("Error during CONNECT: {}", error);
+            LOG.debug("Advice during CONNECT: {}", advice);
+          }
+          if (advice.get("reconnect").equals("handshake")) {
+            LOG.debug("Reconnecting to Salesforce Push Topic");
+            reconnectToTopic();
+          }
+        }
+      });
+
+    bayeuxClient.getChannel(Channel.META_SUBSCRIBE).addListener(
+      (ClientSessionChannel.MessageListener) (channel, message) -> {
+
+        boolean success = message.isSuccessful();
+        if (!success) {
+          String error = (String) message.get("error");
+          if (error != null) {
+            LOG.error("Error during SUBSCRIBE: " + error);
+          }
+        }
+      });
+
+  }
+
+  public void reconnectToTopic() {
+    disconnectStream();
+    createSalesforceListener();
+    waitForHandshake();
+    LOG.debug("start() Client handshake done");
```

Comment:
nit: remove start().
You can move this log line to the end of waitForHandshake(), which avoids duplicating it here and in start()

Reply:
Done. https://github.com/data-integrations/salesforce/commit/547e2e9ff762dc326333289005a82779f0cce293